src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py

```diff
@@ -0,0 +1,521 @@
+              return_raw_logits: bool=False
+              ) -> Union[List[Tuple[Heatmap, Heatmap]], Tuple[Heatmap, Heatmap], np.ndarray]:
+        """
+        This method performs inference on the batch provided.
+
+        :param batch: Object that holds a batch of data to run inference on
+        :type batch: Image class type or list of Image class type
+        :param return_raw_logits: Whether the output should be transformed into the OpenDR target class.
+        :type return_raw_logits: bool
+        :return: A list of predicted targets
+        :rtype: list of tuples of Heatmap class type or list of numpy arrays
+        """
+        if self.model is None:
+            raise RuntimeError('No model loaded.')
+        if not self._is_model_trained:
+            warnings.warn('The current model has not been trained.')
+        self.model.eval()
+
+        # Build the data pipeline
+        test_pipeline = Compose(self._cfg.test_pipeline[1:])
+        device = next(self.model.parameters()).device
+
+        # Convert to the format expected by the mmdetection API
+        single_image_mode = False
+        if isinstance(batch, Image) or isinstance(batch, ImageWithFilename):
+            batch = [batch]
+            single_image_mode = True
+        mmdet_batch = []
+        for img in batch:
+            if isinstance(img, ImageWithFilename):
+                filename = img.filename
+            else:
+                filename = None
+            mmdet_img = {'filename': filename, 'img': img.numpy(), 'img_shape': img.numpy().shape,
+                         'ori_shape': img.numpy().shape}
+            mmdet_img = test_pipeline(mmdet_img)
+            mmdet_batch.append(scatter(collate([mmdet_img], samples_per_gpu=1), [device])[0])
+
+        results = []
+        with torch.no_grad():
+            for data in mmdet_batch:
+                data['eval'] = 'panoptic'
+                prediction = self.model(return_loss=False, rescale=True, **data)[0]
+
+                if return_raw_logits:
+                    results.append(prediction)
+                else:
+                    instance_pred, category_pred, _ = prediction
+                    instance_pred = instance_pred.numpy()
+                    semantic_pred = category_pred[instance_pred].numpy()
+
+                    # Some pixels have not gotten a semantic class assigned because they are marked as stuff by the
+                    # instance head but not by the semantic segmentation head
+                    # We mask them as 255 in the semantic segmentation map
+
+                    instance_pred = Heatmap(instance_pred.astype(np.uint8))
+                    semantic_pred = Heatmap(semantic_pred.astype(np.uint8))
+                    results.append((instance_pred, semantic_pred))
+
+        if single_image_mode:
+            return results[0]
+        else:
+            return results
+
+    def save(self, path: str) -> bool:
```

Comment:
There is a specific OpenDR format for saving/loading models that encapsulates other existing formats in order to provide a uniform interface. This is specified in D2.1 and the already integrated tools provide an example of how to load/save using this format. Essentially, you need to create a folder with the name of your model and include a json metadata file that describes the files that the model needs. Then, you can save the model files inside this folder. Please let me know if you need additional information.

Reply:
Sorry about that. I changed it according to D2.1and also checked other tools' implementations. I do not think that all of them really follow D2.1. It feels also a bit weird that _name_ cannot be specified by the user. I further guess _x_ does not apply in my case.
Perhaps, I am misinterpreting the specification.

**From D2.1**:
The structure of files that are stored for a trained model is the following (assuming that the model is called “name”:):
• “/name/name.json”, which contains the necessary metadata for the model,
• “/name/model_x.pth”, or “/name/model_x.h5tf”, or “/name/model_x.tf”, which stores the actual model. Note that multiple such model files might exist for complex models (this is specified in the metadata JSON file (x corresponds to the id of each model)

Reply:
Should we also update the files uploaded to the OpenDR server or just download the `.pth` file from there and create the `.json` when downloading the model weights?

Reply:
Both options sounds fine.